### ISSUE TITLE
Arbitrary code execution via the _component_ field in an untrusted recipe config

### ISSUE THREAD

**Issue by geo-chen** [pytorch/torchtune#2971]
reported via https://github.com/meta-pytorch/torchtune/security/advisories/GHSA-gw63-xxf2-xxcg on 10 June 2026 - no response so trying here:

### Summary

Loading or validating an untrusted torchtune recipe config (a .yaml data file) executes arbitrary code. A config node names a Python callable in its `_component_` field, and torchtune resolves it by importing the named module and calling it with the config arguments, with no allowlist. Loading a recipe config imports an attacker-named module (running its top-level code) and can call a dangerous callable directly. Recipe configs are explicitly shared and distributed, so loading one is a deliberate action.

### Details

In `torchtune/config/_utils.py`, `_get_component_from_path` (line 39) resolves the `_component_` string with `import_module(module_path)` then `getattr(...)` (single-part branch around line 85, dotted branch around lines 108 to 110). There is no allowlist, no trusted-module set, and no sandbox.

Two reachable sinks:
- `torchtune/config/_instantiate.py:54` `_instantiate_node` calls `_get_component_from_path(obj["_component_"])` then calls the result with kwargs taken from the YAML. Every recipe instantiates attacker-controlled config nodes (for example `recipes/full_finetune_distributed.py` lines 336, 368, 387, 612, 784 for metric_logger, loss, tokenizer, model, optimizer).
- `torchtune/config/_validate.py:28` `validate` (invoked by `tune validate <config>`) calls `_get_component_from_path` for every node. The surprise is that a command documented as only checking that the config is well-formed still imports the attacker-named module, running its top-level code.

So a config with `_component_: os.system` and `command: "..."` calls a dangerous stdlib callable directly, and a config naming any module imports it (running module-body code) even on the nominally read-only validate path. This is the same data-to-code pattern as other config frameworks, shipped with no restriction. No public advisory covers this class on torchtune.

### PoC

```
pip install torchtune torchao
python poc_torchtune_component_rce.py
```

`poc_torchtune_component_rce.py`:

```python
import os, tempfile
from omegaconf import OmegaConf
from torchtune import config

marker = os.path.join(tempfile.gettempdir(), "TT_PWNED")
if os.path.exists(marker): os.remove(marker)

# A DATA file: a shared torchtune fine-tuning recipe (.yaml). No .py shipped.
EVIL = f"""
loss:
  _component_: os.system
  command: "touch {marker}"
"""
cfg = OmegaConf.create(EVIL)
config.instantiate(cfg.loss)   # exactly recipes/full_finetune_distributed.py:368
print("code executed:", os.path.exists(marker))
```

Observed: `config.instantiate(cfg.loss)` runs the command and the marker is created. A `tune validate` variant (a yaml `model: { _component_: evilmod.Trainer }` next to an importable `evilmod.py` with code in its module body) executes the module body during validation, with no instantiation.

### Impact

Arbitrary code execution as the user when a victim loads, validates, or runs an attacker-supplied recipe config. Recipe configs are shared via `tune cp`, the configs registry, and community recipes, so a malicious config is a realistic distribution vector. Running a recipe runs training code by design; the vulnerability is that a data YAML imports and calls an arbitrary module and callable with no allowlist, including on the read-only validate path.

Affected version: 0.6.1 and 0.7.0

**Comment by Solaris-star** [comment]
Opening a fix that allowlists `_component_` import roots to `torchtune` / `torch` / `bitsandbytes` (plus caller globals), blocking paths like `os.system`.
